File: lib/sdk/car/identity/vehicle_capability.dart
/// Vehicle / hardware capabilities — the host's mirror of the SDK's
/// canonical taxonomy in `ilink-sdk/src/types/vehicle-capabilities.ts`.
///
/// **Bit positions are derived from list order — never reorder.** New
/// entries append; removals leave a gap (replace with a tombstone
/// string like `'_removed'` to preserve indices). The SDK's drift
/// check (`scripts/check-capability-drift.mjs`) parses this file
/// looking for `kVehicleCapabilities = <String>[...]`; rename only
/// in coordination with the script.
///
/// Three orthogonal axes the catalog reasons over:
///   * `permissions`           — does the *host* implement the family?
///   * `requiredPermissions`   — is the *publisher* allowed to ship it?
///   * `requiredCapabilities`  — does the *physical car* support it?
///
/// This file owns the third axis on the host.
library;

const List<String> kVehicleCapabilities = <String>[
  'display.read',
  'pkg.read',
  'pkg.launch.ivi',
  'pkg.launch.passenger',
  'pkg.launch.cluster.pixel',
  'pkg.launch.cluster.icons',
  'pkg.launch.dishare',
  'surface.write.ivi',
  'surface.write.passenger',
  'surface.write.cluster',
  'cursor.write',
  'gesture.dispatch',
  'ac.get',
  'ac.set',
  'door.set',
  'window.set',
];

/// Pack a list of capability strings into a single bitmask. Unknown
/// strings are silently skipped — defensive against newer SDKs
/// declaring caps the host doesn't yet know about. The catalog merge
/// uses this on every app per render; keep it allocation-free.
///
/// Named to mirror the SDK's `bitsFromCapabilities`
/// (`ilink-sdk/src/types/vehicle-capabilities.ts`) so the
/// capability-bitmask helpers carry one identifier across TS, Python,
/// and Dart — no per-language alias to drift.
int bitsFromCapabilities(Iterable<String> caps) {
  var bits = 0;
  for (final cap in caps) {
    final idx = kVehicleCapabilities.indexOf(cap);
    if (idx >= 0) bits |= 1 << idx;
  }
  return bits;
}

/// Inverse — turn a bitmask back into the canonical capability list,
/// in taxonomy order (deterministic regardless of how bits were set).
List<String> capabilitiesFromBits(int bits) {
  final out = <String>[];
  for (var i = 0; i < kVehicleCapabilities.length; i++) {
    if ((bits & (1 << i)) != 0) out.add(kVehicleCapabilities[i]);
  }
  return out;
}

/// `app.required ⊆ vehicle.has`. Single bitmask AND, branchless. The
/// catalog filter calls this once per app per render — must stay
/// O(1).
bool hasAllCapabilities(int vehicleBits, int requiredBits) =>
    (vehicleBits & requiredBits) == requiredBits;

File: android/app/src/main/kotlin/com/i99dev/ilink/car/VehicleCapability.kt
package com.i99dev.ilink.car

/**
 * Vehicle / hardware capabilities — host-side mirror of the SDK's
 * canonical taxonomy in `ilink-sdk/src/types/vehicle-capabilities.ts`.
 *
 * **Bit positions are derived from list order — never reorder.** The
 * SDK's drift check parses [ALL] looking for `val ALL: List<String> =
 * listOf(...)`; coordinate any rename with the check script.
 *
 * The host wires the bitmask as a [Long] (64-bit) so we have headroom
 * past the SDK's 31-bit signed-int wire shape; on the wire the value
 * is sent as a [Long] in the [DisplayPlatformPlugin] snapshot.
 */
object VehicleCapability {

    /** Canonical list. Bit position = index. **Append only.** */
    val ALL: List<String> = listOf(
        "display.read",
        "pkg.read",
        "pkg.launch.ivi",
        "pkg.launch.passenger",
        "pkg.launch.cluster.pixel",
        "pkg.launch.cluster.icons",
        "pkg.launch.dishare",
        "surface.write.ivi",
        "surface.write.passenger",
        "surface.write.cluster",
        "cursor.write",
        "gesture.dispatch",
        "ac.get",
        "ac.set",
        "door.set",
        "window.set",
    )

    /** Reverse map for fast lookup. Built once at class init. */
    private val INDEX: Map<String, Int> = ALL.withIndex().associate { (i, c) -> c to i }

    /**
     * Pack a list of capability strings into a [Long] bitmask. Unknown
     * strings are silently skipped (defensive — a future SDK declaring
     * caps an older host doesn't recognise must not crash the snapshot).
     */
    fun bitsOf(caps: Iterable<String>): Long {
        var bits = 0L
        for (cap in caps) {
            val idx = INDEX[cap] ?: continue
            bits = bits or (1L shl idx)
        }
        return bits
    }

    /**
     * Inverse — bitmask to canonical list, in taxonomy order. Used
     * when the host needs to log human-readable names (probe report,
     * Sentry breadcrumbs).
     */
    fun fromBits(bits: Long): List<String> {
        val out = ArrayList<String>()
        for (i in ALL.indices) {
            if ((bits and (1L shl i)) != 0L) out.add(ALL[i])
        }
        return out
    }

    /** `app.required ⊆ vehicle.has`. Single AND. Catalog hot path. */
    fun hasAll(vehicleBits: Long, requiredBits: Long): Boolean =
        (vehicleBits and requiredBits) == requiredBits
}
